File: python/src/threads/team/materialize.py
"""Materialize (spec/schema/README.md, "Teams"; design §4.10): the team worker's first consume of
a starting member. Its prework, outside any transaction, rebinds the member's definition by name;
then one write transaction re-reads the row (gone or no longer starting commits nothing) and opens
the member's branch with thread_started and its task as user_input. A failed rebind is one
complete end-of-member append instead, and no model request is ever made for it. Reference:
spec/tools/fixtures/ops_start.py."""

import json
from collections.abc import Awaitable, Callable, Mapping
from dataclasses import dataclass
from typing import Final, Literal

from pydantic import JsonValue, TypeAdapter
from pydantic.experimental.missing_sentinel import MISSING

from threads.log import (
    BranchId,
    MailEnvelope,
    MemberStartedEvent,
    OperatorSender,
    ParseError,
    ThreadId,
)
from threads.reduce.handlers import to_json
from threads.result import Err, Ok
from threads.store import SqliteStore, Writer
from threads.store.conn import Conn
from threads.store.lease import Lease
from threads.store.lines import Draft, uuid7
from threads.store.opening import BranchOpening
from threads.store.worker import Clock
from threads.team.batch import Batch, Mint
from threads.team.cancel_start import start_cancelled
from threads.team.rows import MemberRow, member_named, pending_to, team_row
from threads.team.settle import AppendContext, SettleContext, settle

type RebindCode = Literal["pin_unavailable", "pin_mismatch", "setup_failed"]


@dataclass(frozen=True, slots=True)
class Rebind:
    """What rebinding the member's definition by name found: ok (with a nested lead's own team,
    which its first append opens), or why not."""

    status: Literal["ok", "pin_unavailable", "pin_mismatch", "setup_failed"]
    team: Mapping[str, JsonValue] | None = None


@dataclass(frozen=True, slots=True)
class _Cancel:
    """A pending cancel: the branch opens and ends cancelled, and the rebind is never tried."""

    mail: MailEnvelope


@dataclass(frozen=True, slots=True)
class Materialized:
    status: Literal["materialized", "not_starting", "cancelled", "rebind_failed"]
    writer: Writer | None = None
    code: RebindCode | None = None


@dataclass(frozen=True, slots=True)
class MaterializeOptions:
    rebind: Callable[[MemberStartedEvent, MailEnvelope], Awaitable[Rebind]]
    """Rebinds the member's definition from its member_started and task (whose sender is the
    starter a dynamic member's block names)."""
    holder: str
    """The lease holder the member's first writer runs under."""
    ttl_ms: int
    clock: Clock
    mint: Mint | None = None
    branch_id: BranchId | None = None
    """The member's branch id; a new one by default."""


@dataclass(frozen=True, slots=True)
class _Starting:
    row: MemberRow
    task: MailEnvelope
    started: MemberStartedEvent


_LINE_ZERO: Final = (
    "agent_name",
    "instructions",
    "model",
    "model_params",
    "adapter",
    "tools",
    "policy",
    "sandbox_provider",
)
"""thread_started takes the pinned config's line-0 fields; the rest of the config is hashed only."""
_OBJECT: TypeAdapter[dict[str, JsonValue]] = TypeAdapter(dict[str, JsonValue])


async def materialize(
    store: SqliteStore, team: str, name: str, o: MaterializeOptions
) -> Ok[Materialized] | Err[ParseError]:
    """Materializes the member `name` of `team`, or reports why nothing was opened."""
    found = await _starting(store, team, name, o.clock())
    if isinstance(found, Err) or found.value is None:
        return found if isinstance(found, Err) else Ok(Materialized("not_starting"))
    s = found.value
    read = await _read(store, s)
    if isinstance(read, Err):
        return read
    text, pinned = read.value
    # A pending cancel ends the member without a rebind, so a setup that can't succeed never
    # stands in its way.
    pending = await store.run(lambda c: pending_to(c, team, name))
    cancel = next((m for m in pending if m.kind == "cancel"), None)
    rebind = _Cancel(cancel) if cancel is not None else await o.rebind(s.started, s.task)
    branch = o.branch_id or BranchId(uuid7(o.clock()))

    def decide(conn: Conn, now: int) -> BranchOpening | None:
        row = member_named(conn, team, name)
        if row is None or row.state != "starting" or row.generation != s.row.generation:
            return None
        batch = Batch(0, now, o.mint)
        _first_events(conn, batch, s, pinned, rebind, (branch, text))
        ttl = o.ttl_ms if isinstance(rebind, Rebind) and rebind.status == "ok" else 0
        held = Lease(o.holder, 1, now + ttl)
        thread = ThreadId(s.started.data.thread_id)
        return BranchOpening("", thread, branch, held, tuple(batch.drafts))

    opened = await store.open_checked(decide, o.clock)
    if isinstance(opened, Err):
        return opened
    return Ok(_outcome(opened.value, rebind))


def _outcome(writer: object, rebind: "Rebind | _Cancel") -> Materialized:
    if not isinstance(writer, Writer):
        return Materialized("not_starting")
    if isinstance(rebind, _Cancel):
        return Materialized("cancelled")
    if rebind.status == "ok":
        return Materialized("materialized", writer)
    return Materialized("rebind_failed", code=rebind.status)


async def _read(
    store: SqliteStore, s: _Starting
) -> Ok[tuple[str, dict[str, JsonValue]]] | Err[ParseError]:
    """What the opening needs from the artifact store: the task's text and the pinned config's
    line-0 fields."""
    text = await _task_text(store, s.task)
    if isinstance(text, Err):
        return text
    config = await store.get_artifact(s.started.data.config_hash)
    if isinstance(config, Err):
        return config
    raw = _OBJECT.validate_python(json.loads(config.value))
    return Ok((text.value, {k: raw[k] for k in _LINE_ZERO if k in raw}))


async def _task_text(store: SqliteStore, task: MailEnvelope) -> Ok[str] | Err[ParseError]:
    """A task's text: inline, or read from the artifact a body over the inline cap became."""
    body = task.body
    if body is MISSING:
        raise AssertionError(f"task {task.mail_id} has no body")
    if body.text is not MISSING:
        return Ok(body.text)
    if body.ref is MISSING:
        raise AssertionError(f"task {task.mail_id} has no text")
    got = await store.get_artifact(body.ref.sha256)
    return got if isinstance(got, Err) else Ok(got.value.decode())


def _first_events(  # noqa: PLR0913, PLR0917 - one opening: where, of whom, with what
    conn: Conn,
    batch: Batch,
    s: _Starting,
    pinned: Mapping[str, JsonValue],
    rebind: "Rebind | _Cancel",
    at: tuple[BranchId, str],
) -> None:
    """thread_started (the pinned config, the member_started's parent) and the task's user_input,
    whose actor is the task's principal; after a failed rebind, the turn closes before any model
    request and the member ends failed with everything an end carries."""
    started = s.started.data
    data: dict[str, JsonValue] = {
        **pinned,
        "config_hash": started.config_hash,
        "parent": to_json(started.parent),
    }
    if isinstance(rebind, Rebind) and rebind.status == "ok" and rebind.team is not None:
        data["team"] = dict(rebind.team)
    batch.add(Draft("thread_started", data))
    task = s.task
    branch, text = at
    actor: dict[str, JsonValue] = {"kind": "host", "principal": to_json(task.provenance.principal)}
    user: dict[str, JsonValue] = {
        "source": "team_task",
        "text": text,
        "mail_id": task.mail_id,
    }
    batch.add(Draft("user_input", user, actor))
    if isinstance(rebind, _Cancel):
        start_cancelled(AppendContext(conn, batch, started.thread_id, branch), task, rebind.mail)
        return
    if rebind.status == "ok":
        return
    code = rebind.status
    batch.add(Draft("turn_completed", {"reason": "error", "code": code}))
    ctx = SettleContext(conn, batch, started.thread_id, branch, to_json(task.provenance), _no_text)
    error: dict[str, JsonValue] = {"code": code, "message": f"rebind failed: {code}"}
    settle(ctx, {"status": "failed", "error": error})


def _no_text(_text: str) -> JsonValue:
    raise AssertionError("a failed rebind's result has no text")


async def _starting(
    store: SqliteStore, team: str, name: str, now: int
) -> Ok[_Starting | None] | Err[ParseError]:
    """The starting row, its pending task and its member_started, read from the starter's log."""
    row = await store.run(lambda c: member_named(c, team, name))
    if row is None or row.state != "starting":
        return Ok(None)
    pending = await store.run(lambda c: pending_to(c, team, name))
    task = next((m for m in pending if m.kind == "task"), None)
    teams = await store.run(lambda c: team_row(c, team))
    if task is None or teams is None:
        return Ok(None)
    started = await started_by(store, (teams.team_log_branch_id, task), row, now)
    return started if isinstance(started, Err) else Ok(_Starting(row, task, started.value))


async def started_by(
    store: SqliteStore, where: tuple[str, MailEnvelope], row: MemberRow, now: int
) -> Ok[MemberStartedEvent] | Err[ParseError]:
    """A member's member_started, read from its starter's log: the team log (`where`'s branch)
    for an operator start, else the log the task (`where`'s envelope) was sent from."""
    team_log, task = where
    if isinstance(task.from_, OperatorSender):
        starter: Ok[BranchId] | Err[ParseError] = Ok(BranchId(team_log))
    else:
        starter = await store.root(ThreadId(task.causal.thread_id))
    if isinstance(starter, Err):
        return starter
    log = await store.read(starter.value, now)
    if isinstance(log, Err):
        return log
    started = next(
        (
            e
            for e in log.value.fold.events
            if isinstance(e, MemberStartedEvent)
            and e.data.member.name == row.name
            and e.data.member.generation == row.generation
        ),
        None,
    )
    if started is None:
        return Err(ParseError("log_corrupt", f"no member_started for {row.name}"))
    return Ok(started)

File: typescript/packages/core/src/team/materialize.ts
import { z } from "zod";
import type { EventOf } from "../fold/state";
import {
  BranchId,
  type MailEnvelope,
  type TeamSettings,
  ThreadStartedData,
} from "../log";
import { knownEvents } from "../reduce";
import { err, ok, type Result } from "../result";
import type { LogStore, Writer } from "../store";
import type { ArtifactStore } from "../store/artifacts";
import { READ_ONLY, type Tx } from "../store/driver";
import { uuidv7 } from "../store/encode";
import { type LogError, logError } from "../verify/error";
import { Batch, MINT, type Mint } from "./batch";
import type { DynamicChoice } from "./dynamic";
import {
  type MemberRow,
  mailEnvelope,
  memberNamed,
  pendingTo,
  teamRow,
} from "./rows";
import { settle } from "./settle";

// Materialize (spec/schema/README.md, "Teams"; design §4.10): the team worker's first consume of
// a starting member. Its prework, outside any transaction, rebinds the member's definition by
// name; then one write transaction re-reads the row (gone or no longer starting commits nothing)
// and opens the member's branch with thread_started and its task as user_input. A failed rebind
// is one complete end-of-member append instead, and no model request is ever made for it.
// Reference: spec/tools/fixtures/ops_start.py.

/** What rebinding the member's definition by name found. */
export type Rebind =
  | {
      readonly status: "ok";
      /** A nested lead's own team, which its first append opens. */
      readonly team?: z.input<typeof TeamSettings>;
    }
  | { readonly status: "pin_unavailable" | "pin_mismatch" };

export type Materialized =
  | { readonly status: "materialized"; readonly writer: Writer }
  | { readonly status: "not_starting" }
  | {
      readonly status: "rebind_failed";
      readonly code: "pin_unavailable" | "pin_mismatch";
    };

export type MaterializeOptions = {
  readonly artifacts: ArtifactStore;
  /** Rebinds by name; a dynamic agent's member with the choice its starter recorded. */
  readonly rebind: (
    agent: string,
    configHash: string,
    choice: DynamicChoice | undefined,
  ) => Promise<Rebind>;
  /** The lease holder the member's first writer runs under. */
  readonly holder: string;
  readonly ttlMs: number;
  readonly mint?: Mint;
  /** The member's branch id; a new one by default. */
  readonly branchId?: BranchId;
};

/** A starting member: its row, its task and the member_started its starter recorded. */
type Starting = {
  readonly row: MemberRow;
  readonly task: MailEnvelope;
  readonly started: EventOf<"member_started">;
};

// thread_started takes the pinned config's line-0 fields; the rest of the config is hashed only.
const Pinned = z.object(
  ThreadStartedData.pick({
    agent_name: true,
    instructions: true,
    model: true,
    model_params: true,
    adapter: true,
    tools: true,
    policy: true,
    sandbox_provider: true,
  }).shape,
);

/** Materializes the member `name` of `team`, or reports why nothing was opened. */
export async function materialize(
  store: LogStore,
  team: string,
  name: string,
  o: MaterializeOptions,
): Promise<Result<Materialized, LogError>> {
  const found = await starting(store, team, name);
  if (!found.ok) return found;
  if (found.value === undefined) return ok({ status: "not_starting" });
  const { started, task } = found.value;
  const rebind = await o.rebind(
    started.data.agent,
    started.data.config_hash,
    choiceOf(started, task),
  );
  const config = await o.artifacts.get(started.data.config_hash);
  if (!config.ok) return config;
  const pinned = Pinned.parse(
    JSON.parse(new TextDecoder().decode(config.value)),
  );
  const branchId = o.branchId ?? BranchId.parse(uuidv7(store.now()));
  const opened = await store.openBranchChecked(async (tx, now) => {
    const row = await memberNamed(tx, team, name);
    if (
      row?.state !== "starting" ||
      row.generation !== found.value?.row.generation
    )
      return ok(undefined);
    const batch = new Batch(0, now, o.mint ?? MINT);
    await firstEvents(
      tx,
      batch,
      { ...found.value, row },
      pinned,
      rebind,
      branchId,
    );
    return ok({
      threadId: started.data.thread_id,
      branchId,
      lease: {
        holderId: o.holder,
        ttlMs: rebind.status === "ok" ? o.ttlMs : 0,
      },
      drafts: batch.drafts,
    });
  });
  if (!opened.ok) return opened;
  const writer = opened.value;
  if (writer === undefined || writer === "already_open")
    return ok({ status: "not_starting" });
  return rebind.status === "ok"
    ? ok({ status: "materialized", writer })
    : ok({ status: "rebind_failed", code: rebind.status });
}

/**
 * thread_started (the pinned config, the member_started's parent) and the task's user_input,
 * whose actor is the task's principal; after a failed rebind, the turn closes before any model
 * request and the member ends failed with everything an end carries.
 */
async function firstEvents(
  tx: Tx,
  batch: Batch,
  s: Starting,
  pinned: z.infer<typeof Pinned>,
  rebind: Rebind,
  branchId: string,
): Promise<void> {
  const { started, task } = s;
  // start writes a task's body inline.
  const text = task.body?.text;
  if (text === undefined) throw new Error(`task ${task.mail_id} has no text`);
  const nested = rebind.status === "ok" ? rebind.team : undefined;
  batch.add({
    type: "thread_started",
    type_version: 1,
    critical: true,
    actor: { kind: "host" },
    data: {
      ...pinned,
      config_hash: started.data.config_hash,
      parent: started.data.parent,
      ...(nested === undefined ? {} : { team: nested }),
    },
  });
  batch.add({
    type: "user_input",
    type_version: 1,
    critical: true,
    actor: { kind: "host", principal: task.provenance.principal },
    data: {
      source: "team_task",
      text,
      mail_id: task.mail_id,
    },
  });
  if (rebind.status === "ok") return;
  const code = rebind.status;
  batch.add({
    type: "turn_completed",
    type_version: 1,
    critical: true,
    actor: { kind: "host" },
    data: { reason: "error", code },
  });
  await settle(
    {
      tx,
      batch,
      threadId: started.data.thread_id,
      branchId,
      provenance: task.provenance,
      put: async () => {
        throw new Error("a failed rebind's result has no text");
      },
    },
    { status: "failed", error: { code, message: `rebind failed: ${code}` } },
  );
}

/** The starting row, its pending task and its member_started, read from the starter's log. */
async function starting(
  store: LogStore,
  team: string,
  name: string,
): Promise<Result<Starting | undefined, LogError>> {
  const rows = await store.driver.transaction(async (tx) => {
    const row = await memberNamed(tx, team, name);
    if (row?.state !== "starting") return undefined;
    const pending = await pendingTo(tx, team, name);
    const task = pending.find((m) => m.kind === "task");
    return task === undefined ? undefined : { row, task };
  }, READ_ONLY);
  if (rows === undefined) return ok(undefined);
  const { row, task } = rows;
  const started = await startedBy(store, row, task);
  if (!started.ok) return started;
  return started.value === undefined
    ? ok(undefined)
    : ok({ row, task, started: started.value });
}

/**
 * A member's member_started, read from its starter's log (the lead's, or the team log for an
 * operator start) as its task mail names it.
 */
async function startedBy(
  store: LogStore,
  row: MemberRow,
  task: MailEnvelope,
): Promise<Result<EventOf<"member_started"> | undefined, LogError>> {
  const teams = await store.driver.transaction(
    (tx) => teamRow(tx, row.team_id),
    READ_ONLY,
  );
  if (teams === undefined) return ok(undefined);
  const starter =
    "operator" in task.from
      ? ok(teams.team_log_branch_id)
      : await store.mainBranch(task.causal.thread_id);
  if (!starter.ok) return starter;
  const log = await store.read(starter.value);
  if (!log.ok) return log;
  const started = knownEvents(log.value).find(
    (e): e is EventOf<"member_started"> =>
      e.type === "member_started" &&
      e.data.member.name === row.name &&
      e.data.member.generation === row.generation,
  );
  return started === undefined
    ? err(logError("log_corrupt", `no member_started for ${row.name}`))
    : ok(started);
}

/**
 * A running member's recorded choice (every continuation rebinds with it): from its task mail
 * and its starter's member_started. Undefined for a member of a static agent.
 */
export async function recordedChoice(
  store: LogStore,
  row: MemberRow,
  mailId: string | undefined,
): Promise<DynamicChoice | undefined> {
  const task =
    mailId === undefined
      ? undefined
      : await store.driver.transaction(
          (tx) => mailEnvelope(tx, mailId),
          READ_ONLY,
        );
  if (task === undefined) return undefined;
  const started = await startedBy(store, row, task);
  if (!started.ok)
    throw new Error(`member ${row.name}: ${started.error.message}`);
  return started.value === undefined
    ? undefined
    : choiceOf(started.value, task);
}

/**
 * A dynamic member's choice: the define its starter recorded, and who that starter is (the
 * operator, or the lead member its task came from). Undefined for any other member.
 */
function choiceOf(
  started: EventOf<"member_started">,
  task: MailEnvelope,
): DynamicChoice | undefined {
  const { define } = started.data;
  if (define === undefined) return undefined;
  return {
    define,
    starter: "operator" in task.from ? "operator" : task.from.name,
  };
}
